src/sema/unused_variable.rs

```diff
@@ -362,21 +362,26 @@ pub fn emit_warning_local_variable(variable: &symtable::Variable) -> Option<Diag
 
 /// Emit warnings depending on the storage variable usage
 fn emit_warning_contract_variables(variable: &ast::Variable) -> Option<Diagnostic> {
-    if variable.assigned && !variable.read {
-        return Some(generate_unused_warning(
-            variable.loc,
-            &format!(
-                "storage variable '{}' has been assigned, but never read",
-                variable.name
-            ),
-            vec![],
-        ));
-    } else if !variable.assigned && !variable.read {
-        return Some(generate_unused_warning(
-            variable.loc,
-            &format!("storage variable '{}' has never been used", variable.name),
-            vec![],
-        ));
+    if !matches!(
+        variable.visibility,
+        Visibility::Public(_) | Visibility::External(_)
```

Comment:
Why are you restricting the warning? Public variables can also be unused.

Reply:
Any public storage variable becomes an accessor function. So if you have this code:
```
contract c {
      int public foo;
      
      function bar() public returns (int) {
           return 1;
      }
}
```
Then when you compile this, you'll get a contract which you can deploy, and you can call `foo()` on it which returns 0, and you can call `bar()` on it which returns 1.

Reply:
I am detecting the usage of variables accessed by accessor functions. Please check the test `function_call` at `unused_variable_detection.rs`. Even though `foo()` returns zero, it might be never used in any other contract.

Reply:
Please, check also my implementation at `src/unused_variable.rs` line 179.

Reply:
The accessor function can be called in a transaction. You can submit a transaction to the blockchain which calls the foo accessor function, just like any public function (like bar).

Reply:
Makes sense. If that is the case, I cannot remove public storage variables at all, right?

Reply:
Unfortunately not no.